internal/controllers/clusterextension_controller.go

```diff
@@ -112,82 +179,232 @@ func checkForUnexpectedFieldChange(a, b ocv1alpha1.ClusterExtension) bool {
+		return ctrl.Result{}, err
+	}
+
+	switch state {
+	case stateNeedsInstall:
+		rel, err = ac.Install(ext.GetName(), r.ReleaseNamespace, chrt, values, func(install *action.Install) error {
+			install.CreateNamespace = false
+			install.Labels = map[string]string{labels.BundleNameKey: bundle.Name, labels.PackageNameKey: bundle.Package, labels.BundleVersionKey: bundleVersion.String()}
+			return nil
+		}, helmclient.AppendInstallPostRenderer(post))
+		if err != nil {
+			if isResourceNotFoundErr(err) {
+				err = errRequiredResourceNotFound{err}
+			}
+			setInstalledStatusConditionFailed(&ext.Status.Conditions, fmt.Sprintf("%s:%v", ocv1alpha1.ReasonInstallationFailed, err), ext.Generation)
+			return ctrl.Result{}, err
+		}
+	case stateNeedsUpgrade:
+		rel, err = ac.Upgrade(ext.GetName(), r.ReleaseNamespace, chrt, values, helmclient.AppendUpgradePostRenderer(post))
+		if err != nil {
+			if isResourceNotFoundErr(err) {
+				err = errRequiredResourceNotFound{err}
+			}
+			setInstalledStatusConditionFailed(&ext.Status.Conditions, fmt.Sprintf("%s:%v", ocv1alpha1.ReasonUpgradeFailed, err), ext.Generation)
+			return ctrl.Result{}, err
+		}
+	case stateUnchanged:
+		if err := ac.Reconcile(rel); err != nil {
+			if isResourceNotFoundErr(err) {
+				err = errRequiredResourceNotFound{err}
+			}
+			setInstalledStatusConditionFailed(&ext.Status.Conditions, fmt.Sprintf("%s:%v", ocv1alpha1.ReasonResolutionFailed, err), ext.Generation)
+			return ctrl.Result{}, err
+		}
+	default:
+		return ctrl.Result{}, fmt.Errorf("unexpected release state %q", state)
+	}
+
+	relObjects, err := util.ManifestObjects(strings.NewReader(rel.Manifest), fmt.Sprintf("%s-release-manifest", rel.Name))
+	if err != nil {
+		setInstalledStatusConditionFailed(&ext.Status.Conditions, fmt.Sprintf("%s:%v", rukpakv1alpha2.ReasonCreateDynamicWatchFailed, err), ext.Generation)
 		return ctrl.Result{}, err
 	}
 
-	// Let's set the proper Installed condition and InstalledBundle field based on the
-	// existing BundleDeployment object status.
-	mapBDStatusToInstalledCondition(existingTypedBundleDeployment, ext, bundle)
+	for _, obj := range relObjects {
+		uMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+		if err != nil {
+			setInstalledStatusConditionFailed(&ext.Status.Conditions, fmt.Sprintf("%s:%v", rukpakv1alpha2.ReasonCreateDynamicWatchFailed, err), ext.Generation)
+			return ctrl.Result{}, err
+		}
 
-	SetDeprecationStatus(ext, bundle)
+		unstructuredObj := &unstructured.Unstructured{Object: uMap}
+		if err := func() error {
+			r.dynamicWatchMutex.Lock()
+			defer r.dynamicWatchMutex.Unlock()
+
+			_, isWatched := r.dynamicWatchGVKs[unstructuredObj.GroupVersionKind()]
+			if !isWatched {
+				if err := r.controller.Watch(
+					source.Kind(r.cache, unstructuredObj),
+					crhandler.EnqueueRequestForOwner(r.Scheme, r.Client.RESTMapper(), ext, crhandler.OnlyControllerOwner()),
```

Comment:
Yes, likely.
OWNER: @varshaprasad96 too investigate/update.

Reply:
Reconciler implements client interface. So both of them (r.Scheme and r.Client.Scheme()) are the same and shouldn't matter. Going to make rest mapper to be `r.RESTMapper()` instead for consistency.

Reply:
This might have been changed due to the linter...

Reply:
@varshaprasad96 I was thinking more the other way around: it looks like the only reasons schema gets passed into the reconciler struct is because it is required on this line. So you can get rid of this extra field on the reconciler if you get the schema from the client.

But I would say it is a nit and a very low priority.

Reply:
Fixed in https://github.com/operator-framework/operator-controller/pull/887. Both scheme and RESTMapper are from client, so a field on reconciler is no longer required either.